fluent/fluent.go

```diff
@@ -148,27 +176,32 @@ func (f *Fluent) PostWithTime(tag string, tm time.Time, message interface{}) err
 }
 
 func (f *Fluent) EncodeAndPostData(tag string, tm time.Time, message interface{}) error {
-	if data, dumperr := f.EncodeData(tag, tm, message); dumperr != nil {
+	data, dumperr := f.EncodeData(tag, tm, message)
+	if dumperr != nil {
 		return fmt.Errorf("fluent#EncodeAndPostData: can't convert '%s' to msgpack:%s", message, dumperr)
 		// fmt.Println("fluent#Post: can't convert to msgpack:", message, dumperr)
-	} else {
-		f.PostRawData(data)
-		return nil
 	}
+	return f.PostRawDataWithResult(data)
 }
 
 func (f *Fluent) PostRawData(data []byte) {
-	f.mu.Lock()
-	f.pending = append(f.pending, data...)
-	f.mu.Unlock()
-	if err := f.send(); err != nil {
+	if err := f.PostRawDataWithResult(data); err != nil {
+		panic(err)
+	}
+}
+
+func (f *Fluent) PostRawDataWithResult(data []byte) (err error) {
+	buf := make([]byte, len(data))
+	copy(buf, data)
+	postResult := make(chan error)
+	select {
+	case f.postCh <- post{result: postResult, data: buf}:
+		err := <-postResult
+		return err
+	case err = <-f.result:
```

Comment:
I don't know whether this is what you want, I produce an "error" in my case:

1. Shutdown fluentd
2. Post message to fluentd (of course it will fail in L.310), so f.result become a closed chan
3. So each time I post, it will get into L.201, even if I startup fluentd again, nothing can be post after MaxRetry exceeded.

IMO, a "logger" should work even in this case, once the connection recovered, it should try its best to send all logs, even though some log might be dropped since the buffer size is limited.